### PR TITLE
Add PWM

### PR DESCRIPTION
**Pull request by jmeggitt**
Adds basic PWM functions.

### CODE REVIEW

**Comment by jmeggitt** [comment]
Whoops, forgot the docs

**Comment by Lytigas** [comment]
Our goal is to provide a low-level abstractions over the HAL. Since we don't have inheritance we can use composition and/or traits. For now, this means implement a `PWM` struct first, and then implement `PWMSpeedController` on top of it. It appears like you merged the two together. I think this will be confusing for people to use coming from WPILib.

File: wpilib/src/pwm.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use sensor_util;
+use wpilib_sys::*;
+
+pub struct PWM {
```
Comment (Lytigas) [inline wpilib/src/pwm.rs]: I think its a good convention to reorder definitions so intrinsic `impl`s appear directly after the struct definition.

File: wpilib/src/pwm.rs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+impl PWM {
+    #[allow(clippy::new_ret_no_self)]
+    pub fn new(channel: i32) -> HalResult<Self> {
+        if !sensor_util::check_pwm_channel(channel) {
+            return Err(HalError(0));
+        }
+
+        let handle = hal_call!(HAL_InitializePWMPort(HAL_GetPort(channel)))?;
+
+        hal_call!(HAL_SetPWMDisabled(handle));
```
Comment (Lytigas) [inline wpilib/src/pwm.rs]: This line and the next need a `?` to fail the constructor if an error occurs.

File: wpilib/src/pwm.rs [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    /// Get the channel of this device.
+    pub fn get_channel(&self) -> i32 {
+        self.channel
+    }
+}
+
+impl Drop for PWM {
+    fn drop(&mut self) {
+        let _ = hal_call!(HAL_SetPWMDisabled(self.handle));
```
Comment (Lytigas) [inline wpilib/src/pwm.rs]: Why the `let _ = `, I don't think its necessary.

Comment (jmeggitt) [reply wpilib/src/pwm.rs]: I assumed it was after seeing it in other places.
https://github.com/Lytigas/first-rust-competition/blob/d36567493b39d4e06d78108948637240886a3235/wpilib/src/ds.rs#L363

Comment (Lytigas) [reply wpilib/src/pwm.rs]: Also add [`.ok()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) to explicitly use the result to be clear about intent.

**Pull request merged**